File: 463IslandPerimeter.py
class Solution(object):
    def islandPerimeter(self, grid):
        """
        :type grid: List[List[int]]
        :rtype: int
        """
        islands = 0
        neighbours = 0

        for i in range(len(grid)):
        	for j in range(len(grid[i])):
        		if grid[i][j] == 1:
        			islands += 1
        			if i+1 < len(grid) and grid[i+1][j] == 1:
        				neighbours += 1
        			if j+1 < len(grid[i]) and grid[i][j+1] == 1:
        				neighbours += 1

        return islands * 4 - neighbours * 2

File: 463IslandPerimeter.cpp
/*
You are given a map in form of a two-dimensional integer grid where 1 represents land and 0 represents water. 
Grid cells are connected horizontally/vertically (not diagonally). 
The grid is completely surrounded by water, and there is exactly one island (i.e., one or more connected land cells). 
The island doesn't have "lakes" (water inside that isn't connected to the water around the island). 
One cell is a square with side length 1. 
The grid is rectangular, width and height don't exceed 100. Determine the perimeter of the island.

Example:

 [[0,1,0,0],
 [1,1,1,0],
 [0,1,0,0],
 [1,1,0,0]]

Answer: 16
Explanation: The perimeter is the 16 yellow stripes in the image below:
*/

class Solution{
public:
	int islandPerimeter(vector<vector<int>>& gird){
		int ans =0;
		int count = 0;
		int m = grid.size();
		int n = grid[0].size();

		for(int i =0; i < m; ++i){
			for(int j =0; j < n; ++j){
				if(grid[i][j] == 1){
					if(i +1 < m)
						count += 2 * grid[i+1][j];
					if(j+1 < n)
						count += 2 * grid[i][j+1];
					++ans;
				}
			}

		}
		return ans*4 - count;
	}
};
